File: Sem4/MPP_Templates/ZboruriCSharp/FlightServerImpl/FlightServerImpl.cs
using model;
using persistence;
using services;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace flightserver
{
    class FlightServerImpl : MarshalByRefObject, IFlightServer
    {
        private AngajatRepository angajatRepository;
        private IZborRepo zborRepository;
        private IClientRepo clientRepository;
        private IBiletRepo biletRepository;
        private Dictionary<String, IFlightObserver> loggedAngajati;

        public FlightServerImpl(AngajatRepository angajatRepository, IZborRepo zborRepository, IClientRepo clientRepository, IBiletRepo biletRepository)
        {
            this.angajatRepository = angajatRepository;
            this.biletRepository = biletRepository;
            this.clientRepository = clientRepository;
            this.zborRepository = zborRepository;
            loggedAngajati = new Dictionary<string, IFlightObserver>();
        }
        public void CumparaBilet(Bilet bilet, int idZbor, int locuri)
        {
            biletRepository.save(bilet);
            zborRepository.updateZbor(idZbor, locuri);
            notifyAll();
        }

        private void notifyAll()
        {
            IEnumerable<Angajat> angajati = angajatRepository.FindAll();
            foreach (Angajat a in angajati)
            {
                if (loggedAngajati.ContainsKey(a.Id))
                {
                    IFlightObserver observer = loggedAngajati[a.Id];
                    observer.TicketBought(zborRepository.findAll());
                }
            }
        }

        public int FindIndex()
        {
            return biletRepository.findCorrectIndex();
        }

        public IEnumerable<Zbor> GetZboruri()
        {
            IList<Zbor> zborDTOs = new List<Zbor>();
            foreach (Zbor z in zborRepository.findAll())
            {
                if (z.Locuri > 0)
                    zborDTOs.Add(z);
            }
            return zborDTOs;
        }

        public IEnumerable<ZborDTO> GetZboruriCautare(string destinatie, string plecare)
        {
            IList<ZborDTO> zborDTOs = new List<ZborDTO>();
            foreach (Zbor z in zborRepository.FindByDestinatieAndPlecare(destinatie, plecare))
            {
                ZborDTO zborDTO = new ZborDTO(TimeSpan.Parse(z.Plecare.Hour.ToString() + ":" + z.Plecare.Minute.ToString()), z.Locuri);
                zborDTOs.Add(zborDTO);
            }
            return zborDTOs;
        }

        public void LogIn(Angajat angajat, IFlightObserver observer)
        {
            Angajat angajat1 = angajatRepository.LogIn(angajat.Id, angajat.Password);
            if (angajat1 != null)
            {
                if (loggedAngajati.ContainsKey(angajat.Id))
                    throw new FlightException("Employee already logged in!");
                loggedAngajati[angajat.Id] = observer;

            }
            else
                throw new FlightException("Autthetication failed!");
        }

        public void LogOut(Angajat angajat, IFlightObserver observer)
        {
            IFlightObserver localClient = loggedAngajati[angajat.Id];
            if (localClient == null)
                throw new FlightException("Empployee logged out");
            loggedAngajati.Remove(angajat.Id);
        }
        
        Client IFlightServer.FindClient(string numeClient, string adresaClient)
        {
            return clientRepository.findOneByNumeAndAdress(numeClient, adresaClient);
        }

        Zbor IFlightServer.FindZbor(string destinatie, string data, string ora)
        {
            return zborRepository.FindOneByDestinatieAndDateTime(destinatie, data + "T" + ora);
        }
    }
}



/*using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ZboruriCSharp.model;
using ZboruriCSharp.repository;

namespace ZboruriCSharp.service
{
    class Service
    {
        private AngajatRepository repoAngajat;
        private IZborRepo repoZbor;
        private IClientRepo repoClient;
        private IBiletRepo repoBilet;

        public Service(AngajatRepository repoAngajat, IZborRepo repoZbor, IClientRepo repoClient, IBiletRepo repoBilet)
        {
            this.repoAngajat = repoAngajat;
            this.repoBilet = repoBilet;
            this.repoClient = repoClient;
            this.repoZbor = repoZbor;
        }

        public bool LogIn(Angajat angajat)
        {
            return repoAngajat.LogIn(angajat.Id, angajat.Password);
        }

        public IEnumerable<Zbor> GetZboruri()
        {
            IList<Zbor> zborDTOs = new List<Zbor>();
            foreach (Zbor z in repoZbor.findAll())
            {
                if(z.Locuri > 0)
                    zborDTOs.Add(z);
            }
            return zborDTOs;
        }

        public IEnumerable<ZborDTO> GetZboruriCautate(String destinatie, String plecare)
        {
            IList<ZborDTO> zborDTOs = new List<ZborDTO>();
            foreach (Zbor z in repoZbor.FindByDestinatieAndPlecare(destinatie, plecare))
            {
                ZborDTO zborDTO = new ZborDTO(TimeSpan.Parse(z.Plecare.Hour.ToString()+":"+z.Plecare.Minute.ToString()), z.Locuri);
                zborDTOs.Add(zborDTO);
            }
            return zborDTOs;
        }

        public Bilet CumparaBilet(String numeClient, String adresaClient, String turisti, int locuri, String destinatie, String data, String ora)
        {
            Client client = repoClient.findOneByNumeAndAdress(numeClient, adresaClient);
            Zbor zbor = repoZbor.FindOneByDestinatieAndDateTime(destinatie, data + "T" + ora);
            if (zbor.Locuri - locuri > 0)
            {
                int index = repoBilet.findCorrectIndex();
                Bilet bilet = new Bilet(++index, zbor.Id, client.Id, turisti, locuri);
                repoZbor.updateZbor(zbor.Id, zbor.Locuri - locuri);
                repoBilet.save(bilet);    
            }
            return null;
        }
    }
}
*/

File: Sem4/MPP_Templates/Zboruri/FlightServer/src/main/java/server/FlightServerImpl.java
package server;

import flight.domain.*;
import flight.domain.*;
import flight.domain.Angajat;
import flight.domain.Bilet;
import flight.domain.Zbor;
import flight.domain.ZborDTO;
import flight.repository.AngajatRepository;
import flight.repository.BiletRepoInterface;
import flight.repository.ClientRepoInterface;
import flight.repository.ZborRepoInterface;
import services.FlightException;
import services.IFlightObserver;
import services.IFlightServer;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class FlightServerImpl implements IFlightServer{
    private AngajatRepository repoAngajat;
    private ZborRepoInterface repoZbor;
    private ClientRepoInterface repoClient;
    private BiletRepoInterface repoBilet;
    private Map<String, IFlightObserver> loggedAngajati;

    public FlightServerImpl(AngajatRepository repoAngajat, ZborRepoInterface repoZbor, ClientRepoInterface repoClient, BiletRepoInterface repoBilet) {
        this.repoAngajat = repoAngajat;
        this.repoZbor = repoZbor;
        this.repoClient = repoClient;
        this.repoBilet = repoBilet;
        loggedAngajati = new ConcurrentHashMap<>();
    }

    @Override
    public void logIn(Angajat angajat, IFlightObserver observer) throws FlightException {
        Angajat angajatR = repoAngajat.logIn(angajat.getID(), angajat.getPassword());
        if (angajatR != null) {
            if(loggedAngajati.get(angajat.getID())!=null)
                throw new FlightException("Employee already logged in!");
            loggedAngajati.put(angajat.getID(),observer);
        } else
            throw new FlightException("Authentication failed!");
    }

    @Override
    public Iterable<Zbor> getZboruri() {
        Iterable<Zbor> it = repoZbor.findAll();
        List<Zbor> zboruri = new ArrayList<>();
        it.forEach(x -> zboruri.add(x));
        return zboruri.stream()
                .filter(p -> p.getLocuri() > 0).collect(Collectors.toList());
    }

    @Override
    public Iterable<ZborDTO> getZboruriCautare(String destinatie, String plecare) {
        Iterable<Zbor> it = repoZbor.findByDestinatieAndPlecare(destinatie, plecare);
        List<Zbor> zboruri = new ArrayList();
        it.forEach(x -> zboruri.add(x));
        List<ZborDTO> zboruriDTO = new ArrayList<>();
        zboruri.forEach(
                x -> {
                    int hour = x.getPlecare().getHour();
                    int minute = x.getPlecare().getMinute();
                    if (x.getLocuri() > 0)
                        zboruriDTO.add(new ZborDTO(LocalTime.of(hour, minute), x.getLocuri()));
                });
        zboruriDTO.forEach(x -> System.out.println(x));
        return zboruriDTO;
    }

    @Override
    public void zborUpdate(Integer zborId, Integer locuri){
        repoZbor.updateZbor(zborId, locuri);
    }


    @Override
    public Client findClient(String numeClient, String adresaClient) {
        return repoClient.findOneByNumeAndAdress(numeClient, adresaClient);
    }

    @Override
    public Zbor findZbor(String destinatie, String data, String ora) {
        return repoZbor.findOneByDestinatieAndDateTime(destinatie, data + "T" + ora);
    }

    @Override
    public Integer findIndex() {
        return repoBilet.findCorrectIndex();
    }

    @Override
    public void cumparaBilet(Bilet bilet, int id) {
        repoBilet.save(bilet);
        Zbor z = repoZbor.findOne(id);
        repoZbor.updateZbor(id, z.getLocuri()-bilet.getLocuri());
        notifyAllAngajati();
    }

    private void notifyAllAngajati() {
        Iterable<Angajat> angajati=repoAngajat.findAll();
        for(Angajat angajat :angajati){
            IFlightObserver client = loggedAngajati.get(angajat.getID());
            if (client!=null){
                System.out.println("Notifying"+ angajat.getID());
                Iterable<Zbor> zboruri = repoZbor.findAll();
                try {
                    client.ticketBought(zboruri);
                } catch (FlightException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    @Override
    public void logOut(Angajat angajat, IFlightObserver observer) throws FlightException {
        IFlightObserver localClient = loggedAngajati.remove(angajat.getID());
        if(localClient == null)
            throw new FlightException("Employee" + angajat.getID() + "is not logged in");
    }


}
